### DISCUSSION TITLE
How does ViolentMonkey work?

### DISCUSSION THREAD

**Discussion by conanathastings** [violentmonkey/violentmonkey#2135] [General]
I'm working on a small Chrome Extension to run some scripts in various internal pages.

It works fine for hard-coded scripts, but I've started working on the capability to edit scripts (just in-memory for now) and I'm just brought to a halt by CORS errors from Manifest V3.

The last thing I've tried was:

1. Have a `textarea` in a popup where I can edit JavaScript
2. Use chrome.runtime to send/receive the code to my `content.js`
3. Try to eval/new Function/iframe/script-inject

Even with an integrity-hash an injected script will not run.

What is ViolentMonkey's strategy? I've ready a little of the source, but it's a little hard to nail exactly what technique is being used.

Thanks in advance for any help!

**Comment by tophf** [comment]
Violentmonkey is currently using MV2, but for a new extension assuming you want to publish it in the web store, you'll have to use [chrome.userScripts](https://developer.chrome.com/docs/extensions/reference/api/userScripts) API.

Use register() method for scripts that should run automatically.

The execute() method is not yet implemented, so for on-demand execution use `chrome.userScripts.configureWorld({csp: "script-src 'unsafe-eval' 'unsafe-inline' 'self'"})` in the background script, then call register() to declare a controller script that listens to a message and either calls eval() or creates a DOM script element with the code in the message. It will use the CSP you've configured, not the website's. The eval() runs in the isolated world of the userscript, whereas the script element runs in the main world of the page.

**Reply by conanathastings** [reply]
Thank you for the information and the quick reply. I'll give that a try. 👍
